streams/src/test/java/org/apache/kafka/streams/state/internals/InMemorySessionStoreTest.java

```diff
@@ -0,0 +1,451 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import static java.time.Duration.ofMillis;
+
+import static org.apache.kafka.test.StreamsTestUtils.toList;
+import static org.apache.kafka.test.StreamsTestUtils.valuesToList;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.hasItem;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.kafka.clients.producer.MockProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.errors.DefaultProductionExceptionHandler;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.kstream.internals.SessionWindow;
+import org.apache.kafka.streams.processor.internals.MockStreamsMetrics;
+import org.apache.kafka.streams.processor.internals.RecordCollector;
+import org.apache.kafka.streams.processor.internals.RecordCollectorImpl;
+import org.apache.kafka.streams.processor.internals.testutil.LogCaptureAppender;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.SessionStore;
+import org.apache.kafka.streams.state.Stores;
+import org.apache.kafka.test.InternalMockProcessorContext;
+import org.apache.kafka.test.TestUtils;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class InMemorySessionStoreTest {
```

Comment:
missing coverage on:
* expired segments
* retention time
* fetchSession, which doesn't find a session
* fetch

Reply:
Can you clarify:

1) fetch (both fetch methods have unit tests?)
2) what is the difference testing-wise between expired segments and retention time(technically there are no "segments" for this store)? There is a test shouldRemoveExpired()
3) Good thought Re: fetchSession not finding a session. Currently this will cause a NPE, which is true for the equivalent of fetchSession in every store implementation -- none of the other stores test for this. This seems like a bug? I would argue that returning null when finding nothing seems like the appropriate behavior for the underlying stores and the bug is in the deserializing layer, which should check for null before blindly deserializing..WDYT?

Reply:
Sure, I was just going by the "Run InMemorySessionStoreTest with Code Coverage" feature in IDEA. I don't believe in shooting for 100% coverage or anything, I just mentioned the things that looked like you probably wanted to test.

1. In `InMemorySessionStore`, lines 186 and 188 (`fetch(final Bytes key)`), and 193 and 195 (`fetch(final Bytes from, final Bytes to)`) are never invoked.
2. I was just referring to `InMemorySessionStore` lines 105 and 106 (during `put`) as expired records.
For "retention period" I meant line 147, where `fetchSession` returns null. But upon second reading, I see that fetchSession might be returning null for several reasons (which illustrates my point about https://github.com/apache/kafka/pull/6525#discussion_r274714946). Regardless, the fall-though case is not covered.
3. Interesting, yeah, it doesn't seem like this should cause an NPE. I see at least one usage in the code where we check the result for null, so apparently, we do expect this to return null some times. It sounds like a bug to me, either because the deserializer should preserve nulls, or because we should skip the deserializer for nulls (both patterns are present in the code).

Reply:
All three of these are actually unrelated to InMemorySessionStore and require some fixes before they can be addressed/properly tested here -- see https://github.com/apache/kafka/pull/6575

Reply:
Thanks @ableegoldman , are you planning to wait for #6575 to be merged before proceeding with this PR, or just not worry about it right now?

Reply:
I think we should not worry about that here, and add corresponding tests for null fetchSession results in that PR?